torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+  // v2 = b
+  // and variables is [a, b, c], then v2 < v1 because the rightmost depending
+  // variable of v2 is b, and the rightmost depending variable of v1 is c,
+  // and b < c. So in this example, this function will return [v2, v1].
+  // Tensors are always considered as variables and they are always considered
+  // as the rightmost.
+  std::vector<Val*> sortedInputs(const std::list<ValInfo>& variables) {
+    std::unordered_set<Val*> variables_set;
+    variables_set.reserve(variables.size());
+    for (auto v : variables) {
+      variables_set.emplace(v.variable);
+    }
+    std::vector<Val*> sorted_inputs(inputs().begin(), inputs().end());
+    auto compare = [&](Val* v1, Val* v2) {
+      bool v1_lefter_than_v2 = false;
+      auto deps1 = getSubexprDependency(v1, variables_set);
+      auto deps2 = getSubexprDependency(v2, variables_set);
+      if (hasTensor(deps2)) {
+        return true;
+      }
+      if (hasTensor(deps1)) {
+        return false;
+      }
+      for (auto v : variables) {
+        if (deps1.count(v.variable) > 0 && deps2.count(v.variable) == 0) {
+          v1_lefter_than_v2 = false;
+        } else if (
+            deps2.count(v.variable) > 0 && deps1.count(v.variable) == 0) {
+          v1_lefter_than_v2 = true;
+        }
+      }
+      return v1_lefter_than_v2;
+    };
+    std::sort(sorted_inputs.begin(), sorted_inputs.end(), compare);
+    return sorted_inputs;
+  }
+
+ protected:
+  // Add a new value as an input of this expression. If `value` is a constant
+  // scalar, then evalate this constant and merge it with the constant term.
+  // Otherwise, add it as an input.
+  void addInput(Val* value) {
+    TORCH_CHECK(
+        hasSimilarType(dtype(), *value->getDataType()),
+        "Input types should be similar, but got: ",
+        dtype(),
+        ", and ",
+        *value->getDataType());
+    value = foldConstants(value);
+    if (hasValue(value)) {
+      updateConstantTermWith(value);
+    } else {
+      Expr::addInput(value);
+    }
+  }
+
+  // Update the constant term. For example, if getOpType() == "+", and the
+  // current constant term is 8, then updateConstantTermWith(7) will change the
+  // constant term to 15.
+  void updateConstantTermWith(Val* value) {
+    if (!hasConstantTerm()) {
+      addAttribute(value);
+      return;
+    }
+    auto new_constant_term = IrBuilder::newScalar(dtype());
+    IrBuilder::create<BinaryOp>(
+        getOpType(), new_constant_term, getConstantTerm(), value);
+    new_constant_term = foldConstants(new_constant_term);
+    TORCH_INTERNAL_ASSERT(hasValue(new_constant_term));
+    attributes_.at(1) = new_constant_term;
+  }
+};
+
+NVFUSER_DEFINE_CLONE_AND_CREATE(FlattenedAssocCommOp)
+
+// Recursively convert expressions like AddOp(AddOp(a, b), AddOp(c, d)) into
+// FlattenedAdd(a, b, c, d)
+Val* flatten(Val* value) {
+  auto def = value->definition();
+  if (def == nullptr) {
+    return value;
+  }
+  if (isMagicZeroProtectedVariable(value)) {
```

Comment:
Is there any chance this could interfere with index hoisting or generally since what we're attempting to do is prevent the compiler from index hoisting it is indicative we want it to interfere? CC @naoyam

Reply:
The index simplifier will just keep `i0 + nvfuser_zero` untouched as is, but it can move around `(i0 + nvfuser_zero)` as a whole, and in the future, the simplified may also remove it as a whole when it sees things like `(i0 + nvfuser_zero) / 32` where `i0` is in the range `0...31`.

Reply:
To have the magic zero do its job, it's important to not assume the actual value of `nvfuser_zero` is zero, so `(i0 + nvfuser_zero) / 32` should not be removed even when `i0` is in that range.

Also, we make `nvfuser_zero` look like updated at certain points in a kernel by inserting `UPDATE_MAGIC_ZERO`, so hoisting across those update points would also invalidate the point of using the magic zero.

We should definitely remove these constraints if possible and confident, but it isn't the case yet.